File: go/internal/sqlitev1/validator.go
package sqlitev1

import (
	"database/sql"
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"strings"

	_ "modernc.org/sqlite"
)

var (
	// uuidRegex enforces strict UUID shape: lowercase canonical text with accepted version (1-8) and RFC4122/RFC9562-compatible variant (8,9,a,b).
	uuidRegex      = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
	timestampRegex = regexp.MustCompile(`^(0|[1-9][0-9]*)$`)
)

type ValidationResult struct {
	DatabaseUUID string
}

func ValidateReadOnly(path string) (*ValidationResult, error) {
	// Connect strictly read-only
	// Properly escape path for SQLite URI to avoid injection or breaking on '#' / '?'
	parsedURL := &url.URL{
		Scheme: "file",
		Opaque: url.PathEscape(path),
	}
	q := parsedURL.Query()
	q.Set("mode", "ro")
	parsedURL.RawQuery = q.Encode()

	db, err := sql.Open("sqlite", parsedURL.String())
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %w", err)
	}
	defer db.Close()

	// 1. Read PRAGMA application_id
	var appId int64
	err = db.QueryRow("PRAGMA application_id").Scan(&appId)
	if err != nil {
		return nil, fmt.Errorf("failed to read PRAGMA application_id: %w", err)
	}
	if appId != 1447906135 {
		return nil, fmt.Errorf("invalid PRAGMA application_id: %d", appId)
	}

	// 2. Read PRAGMA user_version
	var userVersion int64
	err = db.QueryRow("PRAGMA user_version").Scan(&userVersion)
	if err != nil {
		return nil, fmt.Errorf("failed to read PRAGMA user_version: %w", err)
	}
	if userVersion != 1 {
		return nil, fmt.Errorf("invalid PRAGMA user_version: %d", userVersion)
	}

	// 3. Ensure storage_metadata_tbl exists and fetch all properties
	// In SQLite, checking if table exists:
	var tableName string
	err = db.QueryRow("SELECT name FROM sqlite_master WHERE type='table' AND name='storage_metadata_tbl'").Scan(&tableName)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, errors.New("missing storage_metadata_tbl")
		}
		return nil, fmt.Errorf("failed to query storage_metadata_tbl existence: %w", err)
	}

	rows, err := db.Query("SELECT property, value FROM storage_metadata_tbl")
	if err != nil {
		return nil, fmt.Errorf("failed to query storage_metadata_tbl: %w", err)
	}
	defer rows.Close()

	metadata := make(map[string]string)
	for rows.Next() {
		var prop, val string
		if err := rows.Scan(&prop, &val); err != nil {
			return nil, fmt.Errorf("failed to scan metadata row: %w", err)
		}
		metadata[prop] = val
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("metadata rows error: %w", err)
	}

	// 4. Validate metadata fields
	if val, ok := metadata["storage_format_id"]; !ok || val != "vault.moukaeritai.work.storage" {
		return nil, fmt.Errorf("invalid storage_format_id: %v", metadata["storage_format_id"])
	}
	if val, ok := metadata["format_major"]; !ok || val != "1" {
		return nil, fmt.Errorf("invalid format_major: %v", metadata["format_major"])
	}
	if val, ok := metadata["format_minor"]; !ok || val != "0" {
		return nil, fmt.Errorf("invalid format_minor: %v", metadata["format_minor"])
	}
	if val, ok := metadata["schema_version"]; !ok || val != "1" {
		return nil, fmt.Errorf("invalid schema_version: %v", metadata["schema_version"])
	}
	if val, ok := metadata["required_features"]; !ok || val != "[]" {
		return nil, fmt.Errorf("invalid required_features: %v", metadata["required_features"])
	}
	if val, ok := metadata["optional_features"]; !ok || val != "[]" {
		return nil, fmt.Errorf("invalid optional_features: %v", metadata["optional_features"])
	}

	dbUuid, ok := metadata["database_uuid"]
	if !ok || !uuidRegex.MatchString(dbUuid) {
		return nil, fmt.Errorf("invalid database_uuid: %v", metadata["database_uuid"])
	}

	createdAt, ok := metadata["created_at_ms"]
	if !ok || !timestampRegex.MatchString(createdAt) {
		return nil, fmt.Errorf("invalid created_at_ms: %v", metadata["created_at_ms"])
	}

	createdByLib, ok := metadata["created_by_library"]
	if !ok || strings.TrimSpace(createdByLib) == "" {
		return nil, fmt.Errorf("invalid created_by_library")
	}
	createdByVer, ok := metadata["created_by_version"]
	if !ok || strings.TrimSpace(createdByVer) == "" {
		return nil, fmt.Errorf("invalid created_by_version")
	}

	// PRAGMA matches against metadata
	if val, ok := metadata["sqlite_application_id"]; !ok || val != "1447906135" {
		return nil, fmt.Errorf("invalid or missing metadata sqlite_application_id: %v", val)
	}
	if val, ok := metadata["sqlite_user_version"]; !ok || val != "1" {
		return nil, fmt.Errorf("invalid or missing metadata sqlite_user_version: %v", val)
	}

	return &ValidationResult{
		DatabaseUUID: dbUuid,
	}, nil
}

File: zig/src/validator.zig
const std = @import("std");
const sqlite = @import("sqlite.zig");
const uuid = @import("uuid.zig");

pub const ValidationError = error{
    MissingMetadataTable,
    InvalidPragmaApplicationId,
    InvalidPragmaUserVersion,
    MissingMetadataKey,
    InvalidMetadataValue,
    InvalidUuid,
    InvalidTimestamp,
} || sqlite.SQLiteError || uuid.UuidError || std.fmt.ParseIntError;

pub fn validateReadOnly(path: [:0]const u8) !void {
    var db = try sqlite.Database.openReadOnly(path);
    defer db.close();

    try db.exec("PRAGMA foreign_keys = ON");

    // Check application_id
    const app_id = try db.queryPragmaInt("PRAGMA application_id");
    if (app_id != 1447906135) {
        return ValidationError.InvalidPragmaApplicationId;
    }

    // Check user_version
    const user_version = try db.queryPragmaInt("PRAGMA user_version");
    if (user_version != 1) {
        return ValidationError.InvalidPragmaUserVersion;
    }

    // Check table exists
    if (!try db.checkTableExists("storage_metadata_tbl")) {
        return ValidationError.MissingMetadataTable;
    }

    // Fetch and validate metadata
    var stmt = try sqlite.Statement.prepare(&db, "SELECT property, value FROM storage_metadata_tbl");
    defer stmt.finalize();


    var allocator = std.heap.page_allocator;
    var metadata = std.StringHashMap([]const u8).init(allocator);
    defer {
        var it = metadata.iterator();
        while (it.next()) |entry| {
            allocator.free(entry.key_ptr.*);
            allocator.free(entry.value_ptr.*);
        }
        metadata.deinit();
    }

    while (try stmt.step()) {
        const prop = stmt.columnText(0);
        const val = stmt.columnText(1);
        if (prop != null and val != null) {
            const dup_prop = try allocator.dupe(u8, prop.?);
            // Handle error in second dupe correctly by freeing the first
            const dup_val = allocator.dupe(u8, val.?) catch |err| {
                allocator.free(dup_prop);
                return err;
            };
            metadata.put(dup_prop, dup_val) catch |err| {
                allocator.free(dup_prop);
                allocator.free(dup_val);
                return err;
            };

        }
    }

    // Validate specific metadata
    try expectMetadataValue(&metadata, "storage_format_id", "vault.moukaeritai.work.storage");
    try expectMetadataValue(&metadata, "format_major", "1");
    try expectMetadataValue(&metadata, "format_minor", "0");
    try expectMetadataValue(&metadata, "schema_version", "1");
    try expectMetadataValue(&metadata, "required_features", "[]");
    try expectMetadataValue(&metadata, "optional_features", "[]");
    try expectMetadataValue(&metadata, "sqlite_application_id", "1447906135");
    try expectMetadataValue(&metadata, "sqlite_user_version", "1");

    // UUID
    if (metadata.get("database_uuid")) |val| {
        uuid.validateCanonical(val) catch return ValidationError.InvalidUuid;
    } else {
        return ValidationError.MissingMetadataKey;
    }

    // Timestamps
    if (metadata.get("created_at_ms")) |val| {
        if (!isCanonicalTimestamp(val)) return ValidationError.InvalidTimestamp;
    } else {
        return ValidationError.MissingMetadataKey;
    }

    // Created by
    if (metadata.get("created_by_library")) |val| {
        if (val.len == 0) return ValidationError.InvalidMetadataValue;
    } else {
        return ValidationError.MissingMetadataKey;
    }

    if (metadata.get("created_by_version")) |val| {
        if (val.len == 0) return ValidationError.InvalidMetadataValue;
    } else {
        return ValidationError.MissingMetadataKey;
    }
}

fn expectMetadataValue(metadata: *std.StringHashMap([]const u8), key: []const u8, expected: []const u8) !void {
    if (metadata.get(key)) |val| {
        if (!std.mem.eql(u8, val, expected)) {
            return ValidationError.InvalidMetadataValue;
        }
    } else {
        return ValidationError.MissingMetadataKey;
    }
}

fn isCanonicalTimestamp(val: []const u8) bool {
    if (val.len == 0) return false;
    if (val.len > 1 and val[0] == '0') return false; // leading zeros not allowed for multi-digit
    for (val) |char| {
        if (char < '0' or char > '9') return false;
    }
    return true;
}
